Python/bytecodes.c

```diff
@@ -2955,18 +2954,35 @@ dummy_func(
             PyCodeObject *code = (PyCodeObject *)func->func_code;
             DEOPT_IF(code->co_argcount != argcount, CALL);
             DEOPT_IF(!_PyThreadState_HasStackSpace(tstate, code->co_framesize), CALL);
+        }
+
+        op(_INIT_CALL_PY_EXACT_ARGS, (method, callable, args[oparg] -- new_frame: _PyInterpreterFrame*)) {
+            int is_meth = method != NULL;
+            int argcount = oparg;
+            if (is_meth) {
+                callable = method;
+                args--;
+                argcount++;
+            }
             STAT_INC(CALL, hit);
-            _PyInterpreterFrame *new_frame = _PyFrame_PushUnchecked(tstate, func, argcount);
+            PyFunctionObject *func = (PyFunctionObject *)callable;
+            new_frame = _PyFrame_PushUnchecked(tstate, func, argcount);
             for (int i = 0; i < argcount; i++) {
                 new_frame->localsplus[i] = args[i];
             }
-            // Manipulate stack directly since we leave using DISPATCH_INLINED().
-            STACK_SHRINK(oparg + 2);
-            SKIP_OVER(INLINE_CACHE_ENTRIES_CALL);
+        }
+
+        op(_PUSH_FRAME, (new_frame: _PyInterpreterFrame* -- unused)) {
```

Comment:
Since `_PUSH_FRAME` is just `frame->return_offset = 0; DISPATCH_INLINED(new_frame);`, it would make sense to spell out `DISPATCH_INLINED`, to clarify which operations that need to be different for tier 1 and tier 2.
Something like:
```C
op(_PUSH_FRAME, (new_frame: _PyInterpreterFrame* -- unused)) {
    frame->return_offset = 0;
    _PyFrame_SetStackPointer(frame, stack_pointer);
    SAVE_IP();   // Equivalent to frame->prev_instr = next_instr - 1;
    new_frame->previous = frame;
    frame = cframe.current_frame = new_frame;
    CALL_STAT_INC(inlined_py_calls);
    if (_Py_EnterRecursivePy(tstate)) {
        goto exit_unwind;
    }
    JUMP_TO(0);  // Jumps to the start of the (new) frame 
    stack_pointer = _PyFrame_GetStackPointer(frame);
}
```
Then we can move the `SAVE_IP` to the start and the `JUMP_TO(0)` to the end and make them their own uops, so that we get:

```
macro(CALL_PY_EXACT_ARGS) =
            unused/1 + // Skip over the counter
            _CHECK_CALL_PY_EXACT_ARGS +
            _INIT_CALL_PY_EXACT_ARGS +
            _SAVE_IP +
            _PUSH_FRAME +
            _JUMP_TO_START;
```

then the code generator only needs to understand `_SAVE_IP` and `_JUMP_TO_START`.

Reply:
For example: https://github.com/python/cpython/commit/cbb0eddcc9daf00dc896c5b092586c8d5e4919e0

In which case the code generators needs to know to push the temporary stack values to the real stack before `_SAVE_FRAME_STATE`, or on seeing `_PyFrame_SetStackPointer`

Reply:
I have to study this more. A problem is that the Tier 1 and Tier 2 versions of `_PUSH_FRAME` are so different. I am working on a mechanism to be able to say
```
#if TIER_ONE
<code for Tier 1>
#else
<code for Tier 2>
#endif
```
I'm not sure yet what you mean with your last remark about pushing temp stack values.

Reply:
Comparing carefully the two versions of `DISPATCH_INLINED` (adding `frame->return_offset = 0` which precedes it in both cases):

In Tier 1:
```
        frame->return_offset = 0;
        assert(tstate->interp->eval_frame == NULL);
        _PyFrame_SetStackPointer(frame, stack_pointer);
        frame->prev_instr = next_instr - 1;
        (NEW_FRAME)->previous = frame;
        frame = cframe.current_frame = (NEW_FRAME);
        CALL_STAT_INC(inlined_py_calls);
        goto start_frame;
```

In Tier 2:
```
        frame->return_offset = 0;
        assert(tstate->interp->eval_frame == NULL);
        _PyFrame_SetStackPointer(frame, stack_pointer);
        frame->prev_instr -= 1;
        (NEW_FRAME)->previous = frame;
        frame = tstate->cframe->current_frame = (NEW_FRAME);
        CALL_STAT_INC(inlined_py_calls);
        stack_pointer = _PyFrame_GetStackPointer(frame);
        ip_offset = (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive;
```

Diff:
```diff
@@ -1,8 +1,9 @@
         frame->return_offset = 0;
         assert(tstate->interp->eval_frame == NULL);
         _PyFrame_SetStackPointer(frame, stack_pointer);
-        frame->prev_instr = next_instr - 1;
+        frame->prev_instr -= 1;
         (NEW_FRAME)->previous = frame;
-        frame = cframe.current_frame = (NEW_FRAME);
+        frame = tstate->cframe->current_frame = (NEW_FRAME);
         CALL_STAT_INC(inlined_py_calls);
-        goto start_frame;
+        stack_pointer = _PyFrame_GetStackPointer(frame);
+        ip_offset = (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive;
```

Note that the Tier 2 version must be preceded by a `SAVE_IP`, which does the equivalent of `frame->prev_instr = next_instr`. If we had a Tier 1 version of `SAVE_IP` we could include it in the macro definition:
```
        macro(CALL_PY_EXACT_ARGS) =
            unused/1 + // Skip over the counter
            _CHECK_PEP_523 +
            _CHECK_FUNCTION_EXACT_ARGS +
            _CHECK_STACK_SPACE +
            _INIT_CALL_PY_EXACT_ARGS +
            SAVE_IP +      // <-------------- added
            _PUSH_FRAME;
```
which would reduce the special-casing in the code generator a bit (it would still need to do something special for `SAVE_IP` to ensure that its `oparg` has the right value, different from the `oparg` of the macro (which is the argument count). This would take care of the first diff chunk (what to assign to `frame->prev_inst`), but it would still be pretty fragile. (Like my current version, it would entice the optimizer to incorrectly try to remove the `SAVE_IP` uop.)

The second diff chunk relates to how we set `cframe.current_frame` -- in Tier 2 we must access this through the `tstate`.

The third and final diff chunk relates to really start using the new frame. In Tier 1, this must actually do the following:
- Check recursion
- Load `stack_pointer`
- Load `next_instr`
- Dispatch to the next opcode.
This is done by the code at `start_frame`.

In Tier 2 there is no `start_frame` label (the only uop that can go to a label is `EXIT_TRACE`, and of course `DEOPT_IF` and `ERROR_IF` also jump). So we load `stack_frame` here. There is no direct equivalent to `next_instr`, but we have to set `ip_offset`, which `SAVE_IP` adds to its `oparg` to get the `prev_instr` value. (This variable is a cache for `frame->code->co_code_adaptive`, to save some memory loads, so whenever `frame` changes we must update it.)

(More later.)

Reply:
There's another thing though, and I think that is what Mark meant. In Tier 1 the code generation for macros is special-cased for `_PUSH_FRAME` so that both the stack adjustment *and* the `next_instr` adjustment are emitted *before* the `_PUSH_FRAME` opcode. This is done so that the flushing of these variables to the frame in the `DISPATCH_INLINED` macro flush the correct values.

But this is really ugly and unprincipled, and the logic is much hairier than the other special cases for `_PUSH_FRAME`. One of Mark's ideas here is to make this special case look for uops using the `SAVE_FRAME_STATE` macro rather than for the specific uop `_PUSH_FRAME`. But detecting when to trigger the special case is only part of the problem -- IMO the worse problem is that the special case itself is so ugly:
```
        dispatch_inlined_special_case = False
        if mgr is managers[-1] and mgr.instr.always_exits.startswith("DISPATCH_INLINED") and mgr.instr.name == "_PUSH_FRAME":
            dispatch_inlined_special_case = True
            temp = mgr.final_offset.clone()
            temp.deeper(StackEffect(UNUSED))  # Hack
            out.stack_adjust(temp.deep, temp.high)
            # Use clone() since adjust_inverse() mutates final_offset
            mgr.adjust_inverse(mgr.final_offset.clone())
            if cache_adjust:
                out.emit(f"next_instr += {cache_adjust};")
```
The last 4 lines here, starting with `# Use clone()`, occur further down too, for the normal case (*after* the final uop). I don't even recall why the `temp.deeper()` call is needed!

I'll mull this over some more.

Reply:
I think I have addressed this. @markshannon Please have another look. Assuming the tests look okay I'll un-draft this.